src/v/cluster/distributed_kv_stm.h

```diff
@@ -0,0 +1,422 @@
+
+ * An easy way to map a key -> coordinator is using a static hash function but
+ * that doesn't give us the flexibility to increase the partition to scale
+ * the operations. For example, we start with 1 partition, realize it is a
+ * bottleneck and want to increase to 3 partitions, the hash function is
+ * stale. To have this flexibility of scaling partitions, we use a routing
+ * partition that tracks the coordinator mappings. It is partition 0 of the
+ * topic that also works as a coordinator along with routing capabilities.
+ *
+ * Routing partition keeps track of repartitioning operations and remembers
+ * the coordinator mappings prior to repartitiong. So any new keys are hashed
+ * across a larger set of partitions while still retaining original mappings.
+ * Migration of keys across coordinators is not supported and can be added
+ * later on.
+ *
+ * The query pattern from clients is expected to be as follows.
+ * - query routing_partition using coordinator_for(key) to get the coordinator
+ *   for key.
+ * - call add_kv(key, val)/delete_kv(key)/value_for(key) on the coordinator.
+ *
+ * For a reference client implementation, look at transforms_offsets topic
+ * that uses this KV store to track consumption offsets for transforms.
+ */
+
+template<class T>
+concept SerdeSerializable = requires(T t, iobuf buf, iobuf_parser parser) {
+    serde::write_async(buf, t);
+    serde::read_async<T>(parser);
+    serde::from_iobuf<T>(std::move(buf));
+    serde::to_iobuf(t);
+};
+
+template<SerdeSerializable Key, SerdeSerializable Value>
+class kv_stm final : public persisted_stm<> {
+public:
+    explicit kv_stm(
+      size_t max_partitions, ss::logger& logger, raft::consensus* raft)
+      : persisted_stm<>("kv_stm.snapshot", logger, raft)
+      , _default_max_partitions(max_partitions)
+      , _is_routing_partition(_raft->ntp().tp.partition == routing_partition){};
+
+    ss::future<> start() override { co_await persisted_stm<>::start(); }
+    ss::future<> stop() override { co_await _gate.close(); }
+
+    ss::future<> apply(const model::record_batch& record_batch) override {
+        if (record_batch.header().type != model::record_batch_type::raft_data) {
+            co_return;
+        }
+        auto holder = _gate.hold();
+        auto units = co_await _lock.hold_read_lock();
+        co_await record_batch.for_each_record_async(
+          [this](model::record r) -> ss::future<> {
+              auto op = reflection::from_iobuf<operation>(r.release_value());
+              if (op.type == op_type::repartitioning) {
+                  do_apply_repartition(std::move(op.data));
+              } else if (op.type == op_type::coordinator_assignment) {
+                  do_apply_assignment(std::move(op.data));
+              } else if (op.type == op_type::kv_data) {
+                  do_apply_offsets(std::move(op.data));
+              } else {
+                  vlog(clusterlog.error, "unknown type: {}", (int)op.type);
```

Comment:
please add a bit more context to this error message f.e. `Unknown kv_stm command type:`